File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque
def solution(maps):
    answer = 0
    rows = len(maps)
    cols = len(maps[0])
    start = []
    end = []
    lever = []
    maps = [list(s) for s in maps]
    for i in range(rows):
        for j in range(cols):
            if maps[i][j] == 'S':
                start += [i,j]
            if maps[i][j] == 'E':
                end += [i,j]
            if maps[i][j] == 'L':
                lever += [i,j]
    
    dx = [1,0,-1,0]
    dy = [0,-1,0,1]
    
    def bfs(s,e):
        s_x, s_y = s[1], s[0]
        e_x, e_y = e[1], e[0]
        visit = [[0] * (cols) for _ in range(rows)]
        q = deque()
        q.append((s_y,s_x))
        answer = -1
        cna = False
        while q:
            y, x = q.popleft()
            if y == e_y and x == e_x:
                answer = visit[y][x]
                can = True
                break
            for i in range(4):
                ny = y + dy[i]
                nx = x + dx[i]
                if 0 <= ny < rows and 0 <= nx < cols:
                    if maps[ny][nx] != 'X':
                        if visit[ny][nx] == 0:
                            visit[ny][nx] = visit[y][x]+1
                            q.append((ny,nx))
    
        return answer

    a = bfs(start,lever)
    b = bfs(lever,end)
    
    if a == -1 or b == -1:
        return -1
    else:
        return a + b
    
        
        
        
        
        
        
    
    
    
    
    # return lever

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;
class Solution {
    public int[][] dir; 
    public int[] start, lever, end;
    public int solution(String[] maps) {
        int answer = -1;
        dir = new int[][]{{1,0},{0,1},{-1,0},{0,-1}};
        for(int i=0; i<maps.length; i++){
            for(int j=0; j<maps[i].length(); j++){
                if(maps[i].charAt(j) == 'S'){
                    start = new int[] {i,j};
                    maps[i] = maps[i].replace('S','O');
                } else if(maps[i].charAt(j) == 'E'){
                    end = new int[] {i,j};
                    maps[i] = maps[i].replace('E','O');
                } else if(maps[i].charAt(j) == 'L'){
                    lever = new int[] {i,j};
                    maps[i] = maps[i].replace('L','O');
                }
            }
        }
        int first = depth(start,lever,maps);
        int second = depth(lever,end,maps);
        if(first != -1 && second != -1){
            answer = first+second;
        }
        return answer;
    }
    public int depth(int[] start, int[] end, String[] maps){
        int row = maps.length;
        int col = maps[0].length();
        int[][] visit = new int[row][col];
        Queue<int[]> q = new ArrayDeque<>();
        q.add(start);
        while(!q.isEmpty()){
            int[] out = q.remove();
            int y = out[0], x = out[1];
            if(y == end[0] && x == end[1]){
                return visit[y][x];
            }
            for(int[] d : dir){
                int ny = y + d[0], nx = x + d[1];
                if(0<=ny && ny<row && 0<=nx && nx<col && maps[ny].charAt(nx) == 'O'){
                    if(visit[ny][nx] == 0){
                        visit[ny][nx] = visit[y][x] + 1;
                        q.add(new int[] {ny,nx});
                        
                    }
                }
            }
        }
        return -1;
    }
}